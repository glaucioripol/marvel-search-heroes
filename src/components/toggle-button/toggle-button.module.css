.toggle_button {
  background-image: url("/assets/toggle_off.svg");
  background-position: center;
  background-size: cover;
  cursor: pointer;
  display: inline-block;
  height: 41px;
  width: 75px;
  transition: background-image 0.05s ease-in-out;
}

.toggle_button.toggle_button--active,
.toggle_button[aria-checked="true"] {
  background-image: url("/assets/toggle_on.svg");
}
