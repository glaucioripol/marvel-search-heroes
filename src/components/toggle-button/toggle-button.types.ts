export type ToggleButtonProperties = Readonly<{
  isChecked: boolean;
  onChange?: (isChecked: boolean) => void;
  tabIndex?: number;
  dataTestId?: string;
}>;
