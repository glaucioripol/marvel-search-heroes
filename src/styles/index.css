:root {
  --default-font: "Roboto", sans-serif;

  /* Colors */
  --red-1: #fdecec;
  --red-2: #fa7c7c;
  --red-3: #fa8787;
  --red-4: #ff0b0b;
  --red-5: #ff1510;
  --red-6: #e92429;

  --grey-1: #404040; /* usando: Search heroes ao lado da logo */
  --grey-2: #464646; /* usando: titulo acima do texto */
  --grey-3: #969696; /* usando: texto acima do input */
  --grey-4: #bdbdbd; /* usando: Número de dados encontrados */

  /* Background colors */
  --background-color: #ffffff;
  --footer-background-color: var(--red-5);

  /* Font sizes */
  /* texto página herói: Quadrinhos e Filmes  */
  --font-x-sm: 18px;
  /* texto padrão */
  --font-sm: 20px;
  /* Descrição do herói */
  --font-md: 24px;
  /* Titulo página inicial */
  --font-xxl: 44px;
  /* Titulo página do herói, input home */
  --font-xl: 36px;
}

/* reset css */
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  background-color: var(--background-color);
  font-family: var(--default-font);
}
