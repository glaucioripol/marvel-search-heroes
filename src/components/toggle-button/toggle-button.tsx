import styles from "./toggle-button.module.css";

import { KeyboardEvent, useCallback, useMemo } from "react";
import { ToggleButtonProperties } from "./toggle-button.types";

const DEFAULT_TAB_INDEX = 0;

export function ToggleButton({
  isChecked,
  onChange,
  tabIndex = DEFAULT_TAB_INDEX,
  dataTestId = "toggle-button-default",
}: ToggleButtonProperties) {
  const [currentTitle, currentClassName] = useMemo(() => {
    const title = isChecked ? "Desativar" : "Ativar";

    const className = [
      styles.toggle_button,
      isChecked ? styles["toggle_button--active"] : "",
    ].join(" ");

    return [title, className];
  }, [isChecked]);

  const handleToggle = useCallback(
    () => void onChange?.(!isChecked),
    [isChecked, onChange],
  );

  const handleKeyPress = useCallback(
    (event: KeyboardEvent<HTMLDivElement>) => {
      event.preventDefault();
      event.stopPropagation();

      const eventsToListen = [" ", "Enter"];

      if (eventsToListen.includes(event.key)) {
        handleToggle();
      }
    },
    [handleToggle],
  );

  return (
    <div
      role="checkbox"
      data-testid={dataTestId}
      aria-labelledby={currentTitle}
      className={currentClassName}
      aria-checked={isChecked}
      aria-label={currentTitle}
      title={currentTitle}
      tabIndex={tabIndex}
      onClick={handleToggle}
      onKeyDown={handleKeyPress}
    />
  );
}
