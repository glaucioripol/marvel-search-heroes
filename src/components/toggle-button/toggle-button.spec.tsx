import {
  defaultRender,
  describe,
  expect,
  fireEvent,
  it,
  vi,
  screen,
  beforeEach,
} from "@/__tests__";
import { ToggleButton } from "./toggle-button";
import styles from "./toggle-button.module.css";

describe("ToggleButton", () => {
  let onChangeMock = vi.fn();

  beforeEach(() => {
    onChangeMock = vi.fn();
  });

  const renderSUT = (checked: boolean) => {
    const utils = defaultRender(
      <ToggleButton isChecked={checked} onChange={onChangeMock} />,
    );

    const button = screen.getByTestId("toggle-button-default");

    return { ...utils, button };
  };

  it("renders without crashing", () => {
    renderSUT(false);
    expect(screen.getByTestId("toggle-button-default")).toBeInTheDocument();
  });

  it("renders with correct initial state", () => {
    const { button, asFragment } = renderSUT(false);
    expect(button).toHaveAttribute("aria-checked", "false");
    expect(button).toHaveAttribute("data-testid", "toggle-button-default");
    expect(button).toHaveAttribute("aria-label", "Ativar");
    expect(button).toHaveAttribute("title", "Ativar");
    expect(button).toHaveClass(styles.toggle_button);
    expect(asFragment()).toMatchSnapshot();
  });

  it("changes state when clicked", () => {
    const { button } = renderSUT(false);
    fireEvent.click(button);
    expect(onChangeMock).toHaveBeenCalledWith(true);
  });

  it("toggles state correctly on repeated clicks", () => {
    const { button } = renderSUT(true);
    fireEvent.click(button);
    expect(onChangeMock).toHaveBeenCalledWith(false);
  });

  it("changes attributes and class when clicked", () => {
    const { button, rerender, asFragment } = renderSUT(false);

    fireEvent.click(button);
    rerender(<ToggleButton isChecked={true} onChange={onChangeMock} />);

    expect(button).toHaveAttribute("aria-checked", "true");
    expect(button).toHaveAttribute("aria-label", "Desativar");
    expect(button).toHaveAttribute("title", "Desativar");
    expect(button).toHaveClass(styles["toggle_button--active"]);
    expect(asFragment()).toMatchSnapshot();
  });

  describe("keyboard interactions", () => {
    it("calls handleToggle on space key press", () => {
      const { button } = renderSUT(false);
      fireEvent.keyDown(button, { key: " " });
      expect(onChangeMock).toHaveBeenCalledWith(true);
    });

    it("calls handleToggle on Enter key press", () => {
      const { button } = renderSUT(false);
      fireEvent.keyDown(button, { key: "Enter" });
      expect(onChangeMock).toHaveBeenCalledWith(true);
    });

    it("does not call handleToggle on other key press", () => {
      const { button } = renderSUT(false);
      fireEvent.keyDown(button, { key: "a" });
      expect(onChangeMock).not.toHaveBeenCalled();
    });
  });
});
