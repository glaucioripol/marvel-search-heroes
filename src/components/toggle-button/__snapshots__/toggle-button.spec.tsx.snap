// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ToggleButton > changes attributes and class when clicked 1`] = `
<DocumentFragment>
  <div
    aria-checked="true"
    aria-label="Desativar"
    aria-labelledby="Desativar"
    class="_toggle_button_7fc16a _toggle_button--active_7fc16a"
    data-testid="toggle-button-default"
    role="checkbox"
    tabindex="0"
    title="Desativar"
  />
</DocumentFragment>
`;

exports[`ToggleButton > renders with correct initial state 1`] = `
<DocumentFragment>
  <div
    aria-checked="false"
    aria-label="Ativar"
    aria-labelledby="Ativar"
    class="_toggle_button_7fc16a "
    data-testid="toggle-button-default"
    role="checkbox"
    tabindex="0"
    title="Ativar"
  />
</DocumentFragment>
`;
